### PR TITLE
VITIS-11275 Add support for telemetry report in XRT core

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
https://jira.xilinx.com/browse/VITIS-11275

Add Telemetry report for future RyzenAI needs.

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
New feature!

#### How problem was solved, alternative solutions (if any) and why they were rejected
Create the required output report, parsing logic, and query requests to display the needed data.

#### Risks (if any) associated the changes in the commit
None.

#### What has been tested and how, request additional testing if necessary
Tested on Windows VM Simulated IPU

All displayed values are simulated and not from the actual firmware.

Report Output
```
Z:\XRT-MCDM\build\WDebug\xilinx\xrt>xbutil examine -d -r telemetry -o test.json --force
------------------------------------------------------------
                        EARLY ACCESS
        This release of xbutil contains early access
         experimental features which may have bugs.
------------------------------------------------------------

---------------------------------
[0000:18:00.1] : RyzenAI-Phoenix
---------------------------------
Telemetry
  L1 Interrupt Count     : 1234

  RTOS Task  Starts  Schedules  Syscalls  DMA Accesses  Resource Acquisitions
  -----------------------------------------------------------------------------
  0          0       3          6         9             12
  1          1       4          7         10            13
  2          2       5          8         11            14

  RTOS DTLBs
  RTOS Task  Region 0 Misses  Region 1 Misses  Region 2 Misses
  --------------------------------------------------------------
  0          0                1                2
  1          3                4                5
  2          6                7                8

  Opcode  Receive Count
  -----------------------
  0       0
  1       1
  2       2
  3       3

  Stream Buffer  Tokens
  -----------------------
  0              0
  1              1

  AIE Column  Deep Sleep Count
  ------------------------------
  0           0
  1           1
  2           2
  3           3
  4           4

Successfully wrote the json file: test.json
```
JSON Output
```
{
    "schema_version": {
        "schema": "JSON",
        "creation_date": "Wed Jan 31 22:51:26 2024 GMT"
    },
    "devices": [
        {
            "interface_type": "pcie",
            "device_id": "0000:18:00.1",
            "device_status": "HEALTHY",
            "telemetry": {
                "level_one_interrupt_count": "1234",
                "rtos_tasks": [
                    {
                        "started_count": "0",
                        "scheduled_count": "3",
                        "syscall_count": "6",
                        "dma_access_count": "9",
                        "resource_acquisition_count": "12",
                        "dtlb_data": [
                            {
                                "dtlb_misses": "0"
                            },
                            {
                                "dtlb_misses": "1"
                            },
                            {
                                "dtlb_misses": "2"
                            }
                        ]
                    },
                    {
                        "started_count": "1",
                        "scheduled_count": "4",
                        "syscall_count": "7",
                        "dma_access_count": "10",
                        "resource_acquisition_count": "13",
                        "dtlb_data": [
                            {
                                "dtlb_misses": "3"
                            },
                            {
                                "dtlb_misses": "4"
                            },
                            {
                                "dtlb_misses": "5"
                            }
                        ]
                    },
                    {
                        "started_count": "2",
                        "scheduled_count": "5",
                        "syscall_count": "8",
                        "dma_access_count": "11",
                        "resource_acquisition_count": "14",
                        "dtlb_data": [
                            {
                                "dtlb_misses": "6"
                            },
                            {
                                "dtlb_misses": "7"
                            },
                            {
                                "dtlb_misses": "8"
                            }
                        ]
                    }
                ],
                "opcodes": [
                    {
                        "received_count": "0"
                    },
                    {
                        "received_count": "1"
                    },
                    {
                        "received_count": "2"
                    },
                    {
                        "received_count": "3"
                    }
                ],
                "stream_buffers": [
                    {
                        "tokens": "0"
                    },
                    {
                        "tokens": "1"
                    }
                ],
                "aie_columns": [
                    {
                        "deep_sleep_count": "0"
                    },
                    {
                        "deep_sleep_count": "1"
                    },
                    {
                        "deep_sleep_count": "2"
                    },
                    {
                        "deep_sleep_count": "3"
                    },
                    {
                        "deep_sleep_count": "4"
                    }
                ]
            }
        }
    ]
}
```

If the proper queries are not implemented
```
Z:\XRT-MCDM\build\WDebug\xilinx\xrt>xbutil examine -d -r telemetry
------------------------------------------------------------
                        EARLY ACCESS
        This release of xbutil contains early access
         experimental features which may have bugs.
------------------------------------------------------------

---------------------------------
[0000:18:00.1] : RyzenAI-Phoenix
---------------------------------
Telemetry
  No telemetry information available

```
#### Documentation impact (if any)
Need to add new report to documentation list for AIE.

### CODE REVIEW

File: src/runtime_src/core/common/query_requests.h [hunk 1]
```diff
@@ -1544,6 +1551,97 @@ struct aie_partition_info : request
...
+    uint64_t misses;
+  };
+
+  struct data {
+    uint64_t context_starts;
+    uint64_t schedules;
+    uint64_t syscalls;
+    uint64_t dma_access;
+    uint64_t resource_acquisition;
+    std::vector<struct dtlb_data> dtlbs;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:1619]:
```suggestion
    std::vector<dtlb_data> dtlbs;
```

File: src/runtime_src/core/common/query_requests.h [hunk 2]
```diff
@@ -1544,6 +1551,97 @@ struct aie_partition_info : request
...
+  };
+
+  using result_type = std::vector<struct data>;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:1622]:
```suggestion
  using result_type = std::vector<data>;
```

File: src/runtime_src/core/common/query_requests.h [hunk 3]
```diff
@@ -1544,6 +1551,97 @@ struct aie_partition_info : request
...
+// Retrieve the stream buffer telemetry from the device
+// Returns historical data about how the stream buffers operate
+// Applicable to Ryzen devices
+struct stream_buffer_telemetry : request
+{
+  struct data {
+    uint64_t tokens;
+  };
+
+  using result_type = std::vector<struct data>;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:1638]:
```suggestion
  using result_type = std::vector<data>;
```

File: src/runtime_src/core/common/query_requests.h [hunk 4]
```diff
@@ -1544,6 +1551,97 @@ struct aie_partition_info : request
...
+// While the AIE status is for live information. This is historical information
+// of the AIE column operation.
+// This query is available for Ryzen devices
+struct aie_telemetry : request
+{
+  struct data {
+    uint64_t deep_sleep_count;
+  };
+
+  using result_type = std::vector<struct data>;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:1564]:
```suggestion
  using result_type = std::vector<data>;
```

File: src/runtime_src/core/common/query_requests.h [hunk 5]
```diff
@@ -1544,6 +1551,97 @@ struct aie_partition_info : request
...
+// Various bits of information are not tied to anything in AIE devices.
+// This is how to get them!
+// This query is available for Ryzen devices
+struct misc_telemetry : request
+{
+  struct data {
+    uint64_t l1_interrupts;
+  };
+
+  using result_type = struct data;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:1581]:
```suggestion
  using result_type = data;
```

File: src/runtime_src/core/common/query_requests.h [hunk 6]
```diff
@@ -1544,6 +1551,97 @@ struct aie_partition_info : request
...
+// Retrieves the opcode telemetry info for the device
+// Opcodes are the commands that are sent to the device such as EXEC_BUF or SYNC_BO
+// This query is available for Ryzen devices
+struct opcode_telemetry : request
+{
+  struct data {
+    uint64_t count;
+  };
+
+  using result_type = std::vector<struct data>;
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:1597]:
```suggestion
  using result_type = std::vector<data>;
```

File: src/runtime_src/core/tools/common/reports/ReportTelemetry.cpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2024 Advanced Micro Devices, Inc. All rights reserved.
+
+#include "ReportTelemetry.h"
+
+#include "core/common/info_telemetry.h"
+#include "tools/common/Table2D.h"
+
+#include <map>
```
Comment (stsoe) [inline src/runtime_src/core/tools/common/reports/ReportTelemetry.cpp:9]: Is std::map used?
```suggestion
```

**Review by stsoe — CHANGES_REQUESTED** [review]
Looks good.  Few suggestions.

File: src/runtime_src/core/tools/common/reports/ReportTelemetry.cpp (hunk 7 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/tools/common/reports/ReportTelemetry.cpp:9]: It is not! Good catch.

**Review by stsoe — APPROVED** [review]

File: src/runtime_src/core/tools/common/reports/ReportTelemetry.cpp [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+void
+ReportTelemetry::writeReport(const xrt_core::device* /*_pDevice*/,
+                             const boost::property_tree::ptree& pt,
+                             const std::vector<std::string>& /*_elementsFilter*/,
+                             std::ostream& _output) const
+{
+  _output << "Telemetry\n";
+
+  boost::property_tree::ptree telemetry_pt = pt.get_child("telemetry");
```
Comment (AShivangi) [inline src/runtime_src/core/tools/common/reports/ReportTelemetry.cpp:205]: Just add the "info not available" but the rest looks good. Thanks for the comments in the query_requests!

**Review by AShivangi — APPROVED** [review]

**Review by chvamshi-xilinx — APPROVED** [review]

**Pull request merged**